File: apps/sreoi-web/src/i18n.ts
/**
 * Arabic and English, with the parts that actually matter for an Arabic reader:
 * direction, Arabic-Indic digits, and real estate vocabulary rather than
 * transliteration.
 */

export type Locale = 'en' | 'ar';

const STRINGS: Record<Locale, Record<string, string>> = {
	en: {
		'brand': 'Saudi Real Estate Opportunity Intelligence',
		'tagline': 'Discover mispriced real estate, not listings',
		'nav.opportunities': 'Opportunities',
		'nav.map': 'Map',
		'filters.heading': 'Filters',
		'filters.district': 'District',
		'filters.type': 'Type',
		'filters.minScore': 'Minimum score',
		'filters.sort': 'Sort by',
		'filters.apply': 'Apply',
		'filters.clear': 'Clear',
		'sort.score': 'Score',
		'sort.discount': 'Discount',
		'sort.newest': 'Newest',
		'sort.confidence': 'Confidence',
		'list.results': 'opportunities',
		'list.empty': 'No opportunities match these filters.',
		'card.score': 'Score',
		'card.discount': 'Discount',
		'card.confidence': 'Confidence',
		'card.area': 'Area',
		'card.capped': 'capped by confidence gate',
		'card.refused': 'Discount refused',
		'synthetic':
			'Demonstration data. Comparable transactions come from a synthetic fixture corpus, not real registered sales. The engine is real; the evidence is generated.',
		'loading': 'Loading…',
		'error': 'Could not reach the API.',
		'signin.heading': 'Sign in',
		'signin.subtitle': 'Every view requires a credential. The organisation you belong to decides which opportunities you can see.',
		'signin.email': 'Email',
		'signin.password': 'Password',
		'signin.submit': 'Sign in',
		'signin.working': 'Signing in…',
		'signin.invalid': 'That email and password did not match. The same answer is given for an unknown address, so this does not confirm whether the account exists.',
		'signin.throttled': 'Too many sign-in attempts from this address. Wait a few minutes and try again.',
		'signin.disabled': 'This deployment authenticates through its identity provider, not a password.',
		'signin.unavailable': 'Could not reach the sign-in service.',
		'nav.signout': 'Sign out',
		'nav.admin': 'Admin',
		'admin.quality': 'Data quality',
		'admin.sources': 'Source health',
		'admin.overall': 'Overall status',
		'admin.opportunities': 'Opportunities',
		'admin.properties': 'Properties',
		'admin.meanConfidence': 'Mean data confidence',
		'admin.completeness': 'Field completeness',
		'admin.flags': 'Quality flags',
		'admin.metric': 'Metric',
		'admin.value': 'Value',
		'admin.thresholds': 'Warn / fail',
		'admin.confidenceDist': 'Data confidence distribution',
		'admin.confidenceNote': 'The gate holds a score below 0.60 at INSUFFICIENT_DATA and caps it below 0.75. A distribution weighted low means thin evidence, not a broken scorer.',
		'admin.verification': 'Verification pass rate by check',
		'admin.provenance': 'Provenance by basis',
		'admin.provenanceNote': 'ACTUAL is observed, RULE is derived from a published rate, ESTIMATE and UNKNOWN are neither. An UNKNOWN material cost line refuses the discount outright.',
		'admin.notApplicable': 'not applicable',
		'admin.source': 'Source',
		'admin.availability': 'Availability',
		'admin.access': 'Legal access',
		'admin.licence': 'Licence',
		'admin.records': 'Records',
		'admin.synthetic': 'synthetic',
		'admin.forbidden': 'These dashboards require the ADMIN role.',
		'detail.back': '← Back to opportunities',
		'detail.assessment': 'Assessment',
		'detail.marketValue': 'Estimated market value',
		'detail.trueCost': 'True acquisition cost',
		'detail.costIncomplete': 'material items unknown',
		'detail.derivation': 'Score derivation — reproducible',
		'detail.derivationNote': 'Same inputs and same versions always produce the identical score. Weights',
		'detail.dimension': 'Dimension',
		'detail.raw': 'Raw',
		'detail.normalised': 'Normalised',
		'detail.weight': 'Weight',
		'detail.contribution': 'Contribution',
		'detail.lineItem': 'Line item',
		'detail.amount': 'Amount',
		'detail.basis': 'Basis',
		'detail.material': 'Material',
		'detail.yes': 'yes',
		'detail.no': 'no',
		'detail.evidence': 'Valuation evidence',
		'detail.comparablesUsed': 'Comparables used',
		'detail.effectiveN': 'effective n',
		'detail.quality': 'Comparable quality',
		'detail.indexTier': 'Index tier',
		'detail.transaction': 'Transaction',
		'detail.adjusted': 'Adjusted SAR/m²',
		'detail.distance': 'Distance (m)',
		'detail.status': 'Status',
		'detail.used': 'used',
		'detail.verification': 'Verification',
		'detail.check': 'Check',
		'detail.finding': 'Finding',
		'detail.timeline': 'Timeline',
		'detail.timelineNote': 'Snapshots are append-only — a price change is a new row, never an overwrite. The sequence is itself the opportunity signal.',
	},
	ar: {
		'brand': 'منصة اكتشاف الفرص العقارية في السعودية',
		'tagline': 'اكتشف العقارات المُسعّرة بأقل من قيمتها، لا مجرد إعلانات',
		'nav.opportunities': 'الفرص',
		'nav.map': 'الخريطة',
		'filters.heading': 'المرشحات',
		'filters.district': 'الحي',
		'filters.type': 'النوع',
		'filters.minScore': 'أدنى درجة',
		'filters.sort': 'الترتيب حسب',
		'filters.apply': 'تطبيق',
		'filters.clear': 'مسح',
		'sort.score': 'الدرجة',
		'sort.discount': 'نسبة الخصم',
		'sort.newest': 'الأحدث',
		'sort.confidence': 'الثقة',
		'list.results': 'فرصة',
		'list.empty': 'لا توجد فرص مطابقة لهذه المرشحات.',
		'card.score': 'الدرجة',
		'card.discount': 'الخصم',
		'card.confidence': 'الثقة',
		'card.area': 'المساحة',
		'card.capped': 'مُقيّدة ببوابة الثقة',
		'card.refused': 'تم رفض احتساب الخصم',
		'synthetic':
			'بيانات تجريبية. الصفقات المقارنة مأخوذة من مجموعة بيانات اصطناعية، وليست صفقات مسجلة حقيقية. المحرك حقيقي، أما الأدلة فمولّدة.',
		'loading': 'جارٍ التحميل…',
		'error': 'تعذّر الوصول إلى الواجهة البرمجية.',
		'signin.heading': 'تسجيل الدخول',
		'signin.subtitle': 'كل صفحة تتطلب بيانات اعتماد. والمؤسسة التي تنتمي إليها تحدد الفرص التي يمكنك رؤيتها.',
		'signin.email': 'البريد الإلكتروني',
		'signin.password': 'كلمة المرور',
		'signin.submit': 'تسجيل الدخول',
		'signin.working': 'جارٍ تسجيل الدخول…',
		'signin.invalid': 'البريد الإلكتروني أو كلمة المرور غير صحيحة. ويُعطى الجواب نفسه لعنوان غير مسجَّل، فهذا لا يؤكد وجود الحساب.',
		'signin.throttled': 'محاولات تسجيل دخول كثيرة من هذا العنوان. انتظر بضع دقائق ثم أعد المحاولة.',
		'signin.disabled': 'يستخدم هذا النشر مزود الهوية للمصادقة، لا كلمة مرور.',
		'signin.unavailable': 'تعذّر الوصول إلى خدمة تسجيل الدخول.',
		'nav.signout': 'تسجيل الخروج',
		'nav.admin': 'الإدارة',
		'admin.quality': 'جودة البيانات',
		'admin.sources': 'صحة المصادر',
		'admin.overall': 'الحالة العامة',
		'admin.opportunities': 'الفرص',
		'admin.properties': 'العقارات',
		'admin.meanConfidence': 'متوسط ثقة البيانات',
		'admin.completeness': 'اكتمال الحقول',
		'admin.flags': 'مؤشرات الجودة',
		'admin.metric': 'المقياس',
		'admin.value': 'القيمة',
		'admin.thresholds': 'تحذير / فشل',
		'admin.confidenceDist': 'توزيع ثقة البيانات',
		'admin.confidenceNote': 'تُبقي البوابة أي درجة دون ٠٫٦٠ عند «بيانات غير كافية»، وتُقيّدها دون ٠٫٧٥. التوزيع المنخفض يعني أدلة ضعيفة، لا خللًا في المُقيّم.',
		'admin.verification': 'نسبة نجاح التحقق حسب الفحص',
		'admin.provenance': 'مصدر البيانات حسب الأساس',
		'admin.provenanceNote': 'ACTUAL مرصود، وRULE مشتق من نسبة منشورة، أما ESTIMATE وUNKNOWN فليسا كذلك. أي بند تكلفة جوهري بأساس UNKNOWN يرفض احتساب الخصم.',
		'admin.notApplicable': 'غير منطبق',
		'admin.source': 'المصدر',
		'admin.availability': 'التوفر',
		'admin.access': 'الوصول النظامي',
		'admin.licence': 'الترخيص',
		'admin.records': 'السجلات',
		'admin.synthetic': 'اصطناعي',
		'admin.forbidden': 'تتطلب هذه اللوحات دور ADMIN.',
		'detail.back': '→ العودة إلى الفرص',
		'detail.assessment': 'التقييم',
		'detail.marketValue': 'القيمة السوقية المقدرة',
		'detail.trueCost': 'التكلفة الفعلية للاقتناء',
		'detail.costIncomplete': 'بنود جوهرية غير معروفة',
		'detail.derivation': 'اشتقاق الدرجة — قابل لإعادة الإنتاج',
		'detail.derivationNote': 'المدخلات نفسها والإصدارات نفسها تنتج دائمًا الدرجة ذاتها. الأوزان',
		'detail.dimension': 'البعد',
		'detail.raw': 'القيمة الخام',
		'detail.normalised': 'المعيارية',
		'detail.weight': 'الوزن',
		'detail.contribution': 'المساهمة',
		'detail.lineItem': 'البند',
		'detail.amount': 'المبلغ',
		'detail.basis': 'الأساس',
		'detail.material': 'جوهري',
		'detail.yes': 'نعم',
		'detail.no': 'لا',
		'detail.evidence': 'أدلة التقييم',
		'detail.comparablesUsed': 'المقارنات المستخدمة',
		'detail.effectiveN': 'العدد الفعلي',
		'detail.quality': 'جودة المقارنات',
		'detail.indexTier': 'مستوى المؤشر',
		'detail.transaction': 'الصفقة',
		'detail.adjusted': 'ريال/م² معدّل',
		'detail.distance': 'المسافة (م)',
		'detail.status': 'الحالة',
		'detail.used': 'مستخدمة',
		'detail.verification': 'التحقق',
		'detail.check': 'الفحص',
		'detail.finding': 'النتيجة',
		'detail.timeline': 'الخط الزمني',
		'detail.timelineNote': 'اللقطات تُضاف ولا تُستبدل — تغيير السعر صف جديد. التسلسل نفسه هو إشارة الفرصة.',
	},
};

export function t(locale: Locale, key: string): string {
	return STRINGS[locale][key] ?? key;
}

export function direction(locale: Locale): 'rtl' | 'ltr' {
	return locale === 'ar' ? 'rtl' : 'ltr';
}

const ARABIC_INDIC = ['٠', '١', '٢', '٣', '٤', '٥', '٦', '٧', '٨', '٩'] as const;

/**
 * Arabic-Indic digits for Arabic. A number rendered in Western digits inside
 * otherwise-Arabic text reads as foreign, and this is a product for Saudi
 * users rather than a translated English one.
 */
export function digits(value: string, locale: Locale): string {
	if (locale !== 'ar') {
		return value;
	}
	return value.replace(/\d/g, d => ARABIC_INDIC[Number(d)] ?? d);
}

export function formatNumber(value: number | null, locale: Locale, fractionDigits = 0): string {
	if (value === null || !Number.isFinite(value)) {
		return '—';
	}
	const formatted = new Intl.NumberFormat(locale === 'ar' ? 'ar-SA' : 'en-US', {
		minimumFractionDigits: fractionDigits,
		maximumFractionDigits: fractionDigits,
	}).format(value);
	return digits(formatted, locale);
}

export function formatPercent(value: number | null, locale: Locale): string {
	if (value === null || !Number.isFinite(value)) {
		return '—';
	}
	return digits(`${(value * 100).toFixed(1)}%`, locale);
}

if (import.meta.vitest != null) {
	describe('i18n', () => {
		it('renders Arabic-Indic digits in Arabic and Western digits in English', () => {
			expect(digits('2026', 'ar')).toBe('٢٠٢٦');
			expect(digits('2026', 'en')).toBe('2026');
		});

		it('sets direction from the locale', () => {
			expect(direction('ar')).toBe('rtl');
			expect(direction('en')).toBe('ltr');
		});

		it('shows an em dash for an absent number rather than 0', () => {
			// An unknown score and a score of zero mean very different things.
			expect(formatNumber(null, 'en')).toBe('—');
			expect(formatPercent(null, 'en')).toBe('—');
		});

		it('formats a percentage from a fraction', () => {
			expect(formatPercent(0.357, 'en')).toBe('35.7%');
		});

		it('falls back to the key rather than rendering undefined', () => {
			expect(t('en', 'no.such.key')).toBe('no.such.key');
		});

		it('has the same keys in both locales, so nothing falls back silently', () => {
			expect(Object.keys(STRINGS.ar).sort()).toEqual(Object.keys(STRINGS.en).sort());
		});
	});
}

File: platform/src/sreoi_api/i18n.py
"""Arabic and English, with real RTL.

Arabic is the primary language of this market, so it is a first-class locale
rather than a translation layer bolted on later: the document direction, the
number formatting and the district names all change with it.
"""

from __future__ import annotations

from typing import Any

DEFAULT_LOCALE = "en"
LOCALES = ("en", "ar")

_ARABIC_INDIC = "٠١٢٣٤٥٦٧٨٩"

CATALOGUE: dict[str, dict[str, str]] = {
    "en": {
        "brand": "Saudi Real Estate Opportunity Intelligence",
        "tagline": "Discover mispriced real estate, not listings",
        "nav.opportunities": "Opportunities",
        "nav.map": "Map",
        "nav.admin": "Sources",
        "nav.api": "API docs",
        "list.title": "Opportunities",
        "list.subtitle": "Ranked by opportunity score. Below 60% data confidence no recommendation is shown at all.",
        "list.empty": "No opportunities match these filters.",
        "list.showing": "showing",
        "list.of": "of",
        "filters.title": "Filters",
        "filters.district": "District",
        "filters.type": "Opportunity type",
        "filters.max_cost": "Max acquisition cost (SAR)",
        "filters.min_discount": "Min discount %",
        "filters.min_score": "Min score",
        "filters.sort": "Sort by",
        "filters.hide_insufficient": "Hide insufficient-data opportunities",
        "filters.apply": "Apply",
        "filters.reset": "Reset",
        "filters.interpreted": "Filters applied",
        "field.cost": "True acquisition cost",
        "field.value": "Estimated market value",
        "field.discount": "Discount",
        "field.ppsqm": "Price / m²",
        "field.confidence": "confidence",
        "field.area": "Area",
        "field.score": "Opportunity score",
        "field.refused": "Refused",
        "detail.assessment": "Assessment",
        "detail.cost": "True acquisition cost",
        "detail.score": "Score derivation — reproducible",
        "detail.evidence": "Valuation evidence",
        "detail.timeline": "Timeline",
        "detail.provenance": "Provenance",
        "detail.no_recommendation": "No recommendation is shown.",
        "detail.comparables_used": "Comparables used",
        "detail.comparable_quality": "Comparable quality",
        "detail.valuation_confidence": "Valuation confidence",
        "detail.index_tier": "Index tier",
        "detail.line_item": "Line item",
        "detail.amount": "Amount",
        "detail.basis": "Basis",
        "detail.material": "Material",
        "detail.total": "Total",
        "detail.unknown": "UNKNOWN",
        "detail.confidence_gaps": "What is holding confidence down",
        "map.title": "Map",
        "map.subtitle": "Draw or pan to filter. District shading is median price per m² from registered transactions.",
        "map.districts": "District median SAR/m²",
        "admin.title": "Source health",
        "admin.subtitle": "A silently dead connector is the failure that matters most.",
        "admin.state": "State",
        "admin.source": "Source",
        "admin.legal": "Legal basis",
        "admin.licence": "Licence",
        "admin.records": "Records",
        "admin.latency": "Latency",
        "admin.last_record": "Last record",
        "admin.checked": "Checked",
        "banner.synthetic": "Demonstration data. Comparable transactions come from a synthetic fixture corpus, not real registered sales. The engine is real; the evidence is generated.",
        "sort.score": "Highest opportunity score",
        "sort.discount": "Largest discount",
        "sort.cost": "Lowest acquisition cost",
        "sort.newest": "Most recently added",
        "otype.AUCTION": "Auction",
        "otype.ASSIGNMENT": "Assignment",
        "otype.RESALE": "Resale",
        "otype.OFF_PLAN_RESALE": "Off-plan resale",
        "otype.DEVELOPER_INVENTORY": "Developer inventory",
        "pclass.APARTMENT": "Apartment",
        "pclass.VILLA": "Villa",
        "pclass.RESIDENTIAL_PLOT": "Residential plot",
        "detail.verification": "Verification",
        "verif.score": "Verification score",
        "verif.internal": "Internal coherence",
        "verif.official": "Official confirmation",
        "verif.check": "Check",
        "verif.status": "Status",
        "verif.class": "Class",
        "verif.finding": "Finding",
        "verif.capped": "Capped",
        "class.EXCEPTIONAL": "Exceptional",
        "class.STRONG": "Strong",
        "class.WORTH_REVIEWING": "Worth reviewing",
        "class.WATCHLIST": "Watchlist",
        "class.WEAK": "Weak opportunity",
        "class.INSUFFICIENT_DATA": "Insufficient data",
    },
    "ar": {
        "brand": "منصة ذكاء الفرص العقارية السعودية",
        "tagline": "اكتشف العقارات المُسعّرة دون قيمتها، لا مجرد الإعلانات",
        "nav.opportunities": "الفرص",
        "nav.map": "الخريطة",
        "nav.admin": "المصادر",
        "nav.api": "توثيق الواجهة",
        "list.title": "الفرص",
        "list.subtitle": "مرتبة حسب درجة الفرصة. عند انخفاض ثقة البيانات دون ٦٠٪ لا تُعرض أي توصية.",
        "list.empty": "لا توجد فرص مطابقة لهذه المرشحات.",
        "list.showing": "عرض",
        "list.of": "من",
        "filters.title": "المرشحات",
        "filters.district": "الحي",
        "filters.type": "نوع الفرصة",
        "filters.max_cost": "أقصى تكلفة استحواذ (ريال)",
        "filters.min_discount": "أدنى خصم ٪",
        "filters.min_score": "أدنى درجة",
        "filters.sort": "الترتيب",
        "filters.hide_insufficient": "إخفاء الفرص ذات البيانات غير الكافية",
        "filters.apply": "تطبيق",
        "filters.reset": "إعادة تعيين",
        "filters.interpreted": "المرشحات المطبقة",
        "field.cost": "التكلفة الفعلية للاستحواذ",
        "field.value": "القيمة السوقية التقديرية",
        "field.discount": "الخصم",
        "field.ppsqm": "السعر / م²",
        "field.confidence": "الثقة",
        "field.area": "المساحة",
        "field.score": "درجة الفرصة",
        "field.refused": "مرفوض",
        "detail.assessment": "التقييم",
        "detail.cost": "التكلفة الفعلية للاستحواذ",
        "detail.score": "اشتقاق الدرجة — قابل لإعادة الإنتاج",
        "detail.evidence": "أدلة التقييم",
        "detail.timeline": "الخط الزمني",
        "detail.provenance": "مصدر البيانات",
        "detail.no_recommendation": "لا تُعرض أي توصية.",
        "detail.comparables_used": "الصفقات المقارنة المستخدمة",
        "detail.comparable_quality": "جودة المقارنات",
        "detail.valuation_confidence": "ثقة التقييم",
        "detail.index_tier": "مستوى المؤشر",
        "detail.line_item": "البند",
        "detail.amount": "المبلغ",
        "detail.basis": "الأساس",
        "detail.material": "جوهري",
        "detail.total": "الإجمالي",
        "detail.unknown": "غير معروف",
        "detail.confidence_gaps": "ما الذي يخفض الثقة",
        "map.title": "الخريطة",
        "map.subtitle": "حرّك الخريطة للتصفية. تظليل الأحياء يمثل وسيط سعر المتر من الصفقات المسجلة.",
        "map.districts": "وسيط سعر المتر للحي",
        "admin.title": "صحة المصادر",
        "admin.subtitle": "المصدر المتوقف بصمت هو الخلل الأخطر.",
        "admin.state": "الحالة",
        "admin.source": "المصدر",
        "admin.legal": "الأساس النظامي",
        "admin.licence": "الترخيص",
        "admin.records": "السجلات",
        "admin.latency": "زمن الاستجابة",
        "admin.last_record": "آخر سجل",
        "admin.checked": "آخر فحص",
        "banner.synthetic": "بيانات توضيحية. الصفقات المقارنة مولّدة اصطناعياً وليست صفقات مسجلة حقيقية. المحرك حقيقي، والأدلة مولّدة.",
        "sort.score": "الأعلى درجة",
        "sort.discount": "الأكبر خصماً",
        "sort.cost": "الأقل تكلفة",
        "sort.newest": "الأحدث إضافة",
        "otype.AUCTION": "مزاد",
        "otype.ASSIGNMENT": "تنازل",
        "otype.RESALE": "إعادة بيع",
        "otype.OFF_PLAN_RESALE": "إعادة بيع على الخارطة",
        "otype.DEVELOPER_INVENTORY": "مخزون المطور",
        "pclass.APARTMENT": "شقة",
        "pclass.VILLA": "فيلا",
        "pclass.RESIDENTIAL_PLOT": "أرض سكنية",
        "detail.verification": "التحقق",
        "verif.score": "درجة التحقق",
        "verif.internal": "الاتساق الداخلي",
        "verif.official": "التأكيد الرسمي",
        "verif.check": "الفحص",
        "verif.status": "الحالة",
        "verif.class": "النوع",
        "verif.finding": "النتيجة",
        "verif.capped": "محدود",
        "class.EXCEPTIONAL": "استثنائية",
        "class.STRONG": "قوية",
        "class.WORTH_REVIEWING": "تستحق المراجعة",
        "class.WATCHLIST": "قائمة المتابعة",
        "class.WEAK": "فرصة ضعيفة",
        "class.INSUFFICIENT_DATA": "بيانات غير كافية",
    },
}


def register_strings(locale: str, mapping: dict[str, str]) -> None:
    """Add translations from a feature module.

    Keeps the catalogue extensible without every feature editing this file.
    Call it at import time from the feature's own module.
    """
    CATALOGUE.setdefault(locale, {}).update(mapping)


def normalise_locale(value: str | None) -> str:
    if value and value.lower()[:2] in LOCALES:
        return value.lower()[:2]
    return DEFAULT_LOCALE


def direction(locale: str) -> str:
    return "rtl" if locale == "ar" else "ltr"


def translator(locale: str) -> Any:
    """Returns t(key) with an English fallback so a missing string is visible, not blank."""
    catalogue = CATALOGUE.get(locale, CATALOGUE[DEFAULT_LOCALE])
    fallback = CATALOGUE[DEFAULT_LOCALE]

    def t(key: str) -> str:
        return catalogue.get(key) or fallback.get(key) or key

    return t


def localise_digits(text: str, locale: str) -> str:
    """Render numerals in Arabic-Indic form for the Arabic locale."""
    if locale != "ar":
        return text
    return text.translate(str.maketrans("0123456789", _ARABIC_INDIC))


def format_number(value: float | int | None, locale: str, decimals: int = 0) -> str:
    if value is None:
        return "—"
    formatted = f"{value:,.{decimals}f}"
    return localise_digits(formatted, locale)
